blobby/blob.py

```diff
@@ -1,248 +0,0 @@
-from typing import BinaryIO, Callable, List, IO, Type, TypeVar, Iterator
-
-import hashlib
-import io
-import logging
-import os
-import tempfile
-import threading
-from contextlib import contextmanager
-from inspect import getmembers, isfunction, getdoc
-
-from mypy_boto3_s3.service_resource import Bucket as S3Bucket
-from botocore.response import StreamingBody
-
-logger = logging.getLogger(__name__)
-
-T = TypeVar('T')
-
-
-class BlobError(Exception):
-    pass
-
-
-def inherit_docstrings(cls: Type[T]) -> Type[T]:
-    for m in getmembers(cls):
-        if isfunction(m) and not m.__doc__:
-            m.__doc__ = getdoc(m)  # pragma: no cover
-
-    return cls
-
-
-@inherit_docstrings
-class BlobReader(io.RawIOBase):
-    """A read-only, file-like object wrapping an S3 blob."""
-    WHENCE_LIST = ', '.join(
-        str(t) for t in (io.SEEK_SET, io.SEEK_CUR, io.SEEK_END)
-    )
-
-    def __init__(self, s3_bucket: S3Bucket, key: str) -> None:
-        self.object = s3_bucket.Object(key=key)
-        self.position = 0
-
-    @contextmanager
-    def as_fifo(self) -> Iterator[BinaryIO]:
-        """Context manager which creates a FIFO and writes the blob to it.
-
-        Returns: an open file object of the read end of the FIFO, which
-            will be closed by the context manager.
-        """
-        with tempfile.TemporaryDirectory() as dir:
-            path = os.path.join(dir, 'fifo')
-            os.mkfifo(path, mode=0o600)
-            writer = threading.Thread(
-                target=self._fifo_writer, args=(path,)
-            )
-            writer.start()
-            with open(path, 'rb') as fd_r:
-                yield fd_r
-                writer.join()
```

Comment:
Are there cases when this would block unexpectedly? e.g. since this is will perform network IO. If so, this should be in the docstring.

Reply:
Actually I think this is just wrong behaviour. We get back here when the context manager is being closed, so the consumer has finished doing whatever it needs to with the FIFO. If it only read the first four bytes of a 1GB blob it should ideally be possible to just continue, which means killing off the thread here rather than just `join`ing it. What `join` ensures is that the context manager within the thread's target is exited properly, therefore cleaning up the FIFO, so we would have to handle that properly.